### ISSUE TITLE
Avoid unnecessary reading when select.rows is specified in read.dta13()

### ISSUE THREAD

**Issue by Kevin-Jin** [sjewo/readstata13#49]
I'm working with a 12GB file with around 4 million rows. Since there's no way I could keep it all in memory, I thought I would be able to process the file in chunks using the select.rows parameter, but I realized that read_dta.cpp actually is inefficiently implemented and reads the entire file in every call no matter what values are passed to select.rows.

I'm currently working on a pull request to implement a fix using fseeko64(). It's fairly trivial since each row is of the same length; the regular string columns are fixed width and the strL columns are simply 64-bit integers, while the doubles, floats, longs, ints, and bytes are all the expected width.

However, I don't actually have a license to Stata so I cannot create a toy data set. The real world data set I'm working with does not make use of every Stata data type. Can someone please create v13 and v14 dta files that have double, float, long, int, byte, and string columns of various lengths, and a few strL columns so that I can test my implementation? It probably would only need 10 rows. Preferably I would also like the equivalent CSV representation so that I can check that the right rows are being skipped.

**Comment by JanMarvin** [comment]
Hello @Kevin-Jin, well chunk-wise reading a file was the reason I implemented this. And yes I know that it reads the full file and returns only the required parts. Better save then sorry was the idea (a while ago it took us some time to fix all the initial fseeks). There already is a data set which has the different types in one of our folders iirc. Let me know if that helps. Looking forward to the pull request

**Comment by JanMarvin** [comment]
@Kevin-Jin I have pushed some code to our fseeko64 branch. For now fseeko64 is used row-wise. I calculate the row-width, check if I need to import it and if not skip the row. I checked with my biggest Stata file (~2GB) and it seems to work quite well. We might want to check if we can skip larger parts e.g., if I want only the first row, skip rlength * (n-1), but this needs some more thinking and I only had a single coffee today. Please let me know what you think.

**Comment by Kevin-Jin** [comment]
[My updates](https://github.com/Kevin-Jin/readstata13/commit/2e5c4a8776e3669e9bcab75cbd17ce87ae89c70d) are less invasive (with less refactoring) but gets the job done with the file I have, at least for non-strL columns. I avoid the loop when skipping rows and instead just multiply row size by the rows skipped when calling fseeko64(). I think it would be a bit cleaner if I just transfer over your use of the named constants in the switch statements into my branch and then push my branch to testing.

**Comment by Kevin-Jin** [comment]
I do like your use of the rlen vector though because it allows us to consider future cases of skipping columns instead of rows. I can transfer that over too if you're fine with using my branch.

**Comment by Kevin-Jin** [comment]
On second thought, I think just updating your branch is fine. Is it possible for you to grant me write access?

**Comment by JanMarvin** [comment]
Sorry please just create a pull request to the branch if you want a modification. Didn't have time to look into your proposal, will do soon

**Comment by JanMarvin** [comment]
I have updated my pull request with your proposal. I like the idea for its simplicity jumping in where required, read whats needed and bail out. This wont allow us to read a sequence of rows, but obviously is the fastest way to read a selected part. And is way prettier code than my initial attempt. I did not copy your warning/stop message. If the dta-file is written incorrectly we will bail at the next `test()` or read nonsense.
